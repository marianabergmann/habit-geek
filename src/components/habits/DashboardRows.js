import React, { Fragment } from "react";

import HabitCheckbox from "./HabitCheckbox";
import "./DashboardRows.scss";

const DashboardRows = ({ habits, toggleDayHabit }) => {
  const dashboardRows = habits.map(
    ({ name, frequency, checked }, containerIndex) => {
      const now = new Date();
      const today = now.getDay() === 0 ? 6 : now.getDay() - 1;

      const daysUntilEndOfWeek = checked[today] ? -today + 6 : -today + 7;
      const checkedDays = checked.filter(checkedDay => checkedDay).length;

      const accomplishedClass = frequency <= checkedDays ? "table-success" : "";
      const failedClass =
        frequency > checkedDays + daysUntilEndOfWeek ? "table-danger" : "";

      return (
        <div
          key={`check-container-${containerIndex}`}
          className="dashboard__row"
        >
          <div className="dashboard__cell-habit-container">
            ({frequency}x) - {name}
          </div>
          <div
            className={`dashboard__cell-frequency-container ${accomplishedClass} ${failedClass}`}
          >
            {[...Array(7).keys()].map(checkIndex => (
              <div
                className="dashboard__cell"
                key={`check-${containerIndex}-${checkIndex}`}
              >
                <HabitCheckbox
                  clickHandler={() => toggleDayHabit(name, checkIndex)}
                  isChecked={checked[checkIndex]}
                  isDisabled={today < checkIndex}
                ></HabitCheckbox>
              </div>
            ))}
          </div>
        </div>
      );
    }
  );

  return <Fragment>{dashboardRows}</Fragment>;
};

export default DashboardRows;
