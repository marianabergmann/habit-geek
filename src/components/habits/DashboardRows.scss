.dashboard__cell .toggle {
  margin: 5px 0 5px 7px;
}
