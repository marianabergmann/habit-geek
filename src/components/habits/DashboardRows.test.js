import React from "react";
import { shallow } from "enzyme";
import DashboardRows from "./DashboardRows";

describe("DashboardRows component", () => {
  it("Renders static elements", () => {
    const dashboardRows = shallow(
      <DashboardRows username="rafa" habits={[]} />
    );
    expect(dashboardRows).toMatchSnapshot();
  });

  it("Shows the correct ammount of habits", () => {
    const checked = [false, false, false, false, false, false, false];
    const dashboardRows = shallow(
      <DashboardRows
        habits={[
          { name: "exercise", frequency: 1, checked },
          { name: "read", frequency: 1, checked }
        ]}
      />
    );
    expect(dashboardRows).toMatchSnapshot();
  });
});
