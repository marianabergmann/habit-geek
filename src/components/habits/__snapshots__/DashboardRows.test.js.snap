// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardRows component Renders static elements 1`] = `<Fragment />`;

exports[`DashboardRows component Shows the correct ammount of habits 1`] = `
<Fragment>
  <div
    className="dashboard__row"
    key="check-container-0"
  >
    <div
      className="dashboard__cell-habit-container"
    >
      (
      1
      x) - 
      exercise
    </div>
    <div
      className="dashboard__cell-frequency-container  "
    >
      <div
        className="dashboard__cell"
        key="check-0-0"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-1"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-2"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-3"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-4"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-5"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-6"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
    </div>
  </div>
  <div
    className="dashboard__row"
    key="check-container-1"
  >
    <div
      className="dashboard__cell-habit-container"
    >
      (
      1
      x) - 
      read
    </div>
    <div
      className="dashboard__cell-frequency-container  "
    >
      <div
        className="dashboard__cell"
        key="check-1-0"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-1"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-2"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-3"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-4"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-5"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-6"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`DashboardRows component marks habits correctly at the last day of the week 1`] = `
<Fragment>
  <div
    className="dashboard__row"
    key="check-container-0"
  >
    <div
      className="dashboard__cell-habit-container"
    >
      (
      6
      x) - 
      ok
    </div>
    <div
      className="dashboard__cell-frequency-container  "
    >
      <div
        className="dashboard__cell"
        key="check-0-0"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-1"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-2"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-3"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-4"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-5"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-6"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
    </div>
  </div>
  <div
    className="dashboard__row"
    key="check-container-1"
  >
    <div
      className="dashboard__cell-habit-container"
    >
      (
      7
      x) - 
      failed
    </div>
    <div
      className="dashboard__cell-frequency-container  table-danger"
    >
      <div
        className="dashboard__cell"
        key="check-1-0"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-1"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-2"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-3"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-4"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-5"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-6"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`DashboardRows component marks habits that are completed for that week 1`] = `
<Fragment>
  <div
    className="dashboard__row"
    key="check-container-0"
  >
    <div
      className="dashboard__cell-habit-container"
    >
      (
      3
      x) - 
      ok
    </div>
    <div
      className="dashboard__cell-frequency-container  "
    >
      <div
        className="dashboard__cell"
        key="check-0-0"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-1"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-2"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-3"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-4"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-5"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-6"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
    </div>
  </div>
  <div
    className="dashboard__row"
    key="check-container-1"
  >
    <div
      className="dashboard__cell-habit-container"
    >
      (
      2
      x) - 
      failed
    </div>
    <div
      className="dashboard__cell-frequency-container table-success "
    >
      <div
        className="dashboard__cell"
        key="check-1-0"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-1"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-2"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-3"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-4"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-5"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-6"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`DashboardRows component marks habits that are not performing well 1`] = `
<Fragment>
  <div
    className="dashboard__row"
    key="check-container-0"
  >
    <div
      className="dashboard__cell-habit-container"
    >
      (
      5
      x) - 
      ok
    </div>
    <div
      className="dashboard__cell-frequency-container  "
    >
      <div
        className="dashboard__cell"
        key="check-0-0"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-1"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-2"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-3"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-4"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-5"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-0-6"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
    </div>
  </div>
  <div
    className="dashboard__row"
    key="check-container-1"
  >
    <div
      className="dashboard__cell-habit-container"
    >
      (
      6
      x) - 
      failed
    </div>
    <div
      className="dashboard__cell-frequency-container  table-danger"
    >
      <div
        className="dashboard__cell"
        key="check-1-0"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-1"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-2"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-3"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-4"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-5"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-1-6"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
    </div>
  </div>
  <div
    className="dashboard__row"
    key="check-container-2"
  >
    <div
      className="dashboard__cell-habit-container"
    >
      (
      7
      x) - 
      failed
    </div>
    <div
      className="dashboard__cell-frequency-container  table-danger"
    >
      <div
        className="dashboard__cell"
        key="check-2-0"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-2-1"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-2-2"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-2-3"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-2-4"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={true}
          isDisabled={false}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-2-5"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
      <div
        className="dashboard__cell"
        key="check-2-6"
      >
        <HabitCheckbox
          clickHandler={[Function]}
          isChecked={false}
          isDisabled={true}
        />
      </div>
    </div>
  </div>
</Fragment>
`;
